data/src/main/java/org/apache/iceberg/MetricsAppender.java

```diff
@@ -0,0 +1,839 @@
+      return fieldResult;
+    }
+
+    @Override
+    public MetricsCollector list(Types.ListType list, MetricsCollector elementResult) {
+      return new MetricsCollectorList(elementResult);
+    }
+
+    @Override
+    public MetricsCollector map(Types.MapType map, MetricsCollector keyResult, MetricsCollector valueResult) {
+      return new MetricsCollectorMap(keyResult, valueResult);
+    }
+
+    @Override
+    public MetricsCollector primitive(Type.PrimitiveType primitive) {
+      int fieldId = stack.peek().fieldId();
+
+      switch (primitive.typeId()) {
+        case BOOLEAN:
+          return new MetricsCollectorBoolean(fieldId, Comparators.forType(primitive));
+        case INTEGER:
+          return new MetricsCollectorInt(fieldId, Comparators.forType(primitive));
+        case LONG:
+          return new MetricsCollectorLong(fieldId, Comparators.forType(primitive));
+        case FLOAT:
+          return new MetricsCollectorFloat(fieldId, Comparators.forType(primitive));
+        case DOUBLE:
+          return new MetricsCollectorDouble(fieldId, Comparators.forType(primitive));
+        case DATE:
+          MetricsCollector metricsCollectorDate = null;
+
+          switch (conf.dateTypeRepresentation) {
+            case INT:
+              metricsCollectorDate = new MetricsCollectorInt(fieldId, Comparators.forType(Types.IntegerType.get()));
+              break;
+            case LOCAL_DATE:
+              metricsCollectorDate = new MetricsCollectorDate(fieldId, Comparators.forType(primitive));
+              break;
+            default:
+              throw new UnsupportedOperationException("Not a supported date type representation: " +
+                  conf.dateTypeRepresentation);
+          }
+
+          return metricsCollectorDate;
+        case TIME:
+          MetricsCollector metricsCollectorTime = null;
+
+          switch (conf.timeTypeRepresentation) {
+            case LOCAL_TIME:
+              metricsCollectorTime = new MetricsCollectorTime(fieldId, Comparators.forType(primitive));
+              break;
+            case LONG:
+              metricsCollectorTime = new MetricsCollectorLong(fieldId, Comparators.forType(Types.LongType.get()));
+              break;
+            default:
+              throw new UnsupportedOperationException("Not a supported time type representation: " +
+                  conf.timeTypeRepresentation);
+          }
+
+          return metricsCollectorTime;
+        case TIMESTAMP:
+          Types.TimestampType timestamp = (Types.TimestampType) primitive;
+          MetricsCollector metricsCollector = null;
+
+          switch (conf.timestampTypeRepresentation) {
+            case LOCAL_DATETIME:
+              metricsCollector = new MetricsCollectorLocalDateTime(fieldId,
+                  Comparators.forType(primitive), timestamp.shouldAdjustToUTC());
+              break;
+            case OFFSET_DATETIME:
```

Comment:
`OffsetDateTime` is used for `TimestampType.withZone()` and `LocalDateTime` is used for `TimestampType.withoutZone()`. Those can coexist in a schema, so it doesn't fit with using a configuration for this. The choice is either longs or Java date/time classes.